sql/core/src/main/scala/org/apache/spark/sql/WindowFunctionDefinition.scala

```diff
@@ -0,0 +1,341 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import scala.language.implicitConversions
+
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.sql.catalyst.expressions._
+
+/**
+ * :: Experimental ::
+ * A set of methods for window function definition for aggregate expressions.
+ * For example:
+ * {{{
+ *   // predefine a window
+ *   val w = partitionBy("name").orderBy("id")
+ *
+ *   df.select(
+ *     first("value")
+ *       over(w).as("first_value"),
+ *     last("value")
+ *       over(w).as("last_value"),
+ *     avg("value")
+ *       over(
+ *       partitionBy("k1")
+ *       .orderBy("k2", "k3")
+ *       .rows
+ *       .following(1)).as("avg_value"),
+ *     max("value")
+ *       .over(
+ *       partitionBy("k2")
+ *       .orderBy("k3")
+ *       .range
+ *       .between
+ *       .preceding(4)
+ *       .and
+ *       .following(3)).as("max_value"))
+ *
+ * }}}
+ *
+ * @param column The bounded the aggregate/window function
+ * @param partitionSpec The partition of the window
+ * @param orderSpec The ordering of the window
+ * @param frame The Window Frame type
+ * @param bindLower A hint of when call the methods `.preceding(n)` `.currentRow()` `.following()`
+ *                  if bindLower == true, then we will set the lower bound, otherwise, we should
+ *                  set the upper bound for the Row/Range Frame.
+ */
+@Experimental
+class WindowFunctionDefinition protected[sql](
+    column: Column = null,
+    partitionSpec: Seq[Expression] = Nil,
+    orderSpec: Seq[SortOrder] = Nil,
+    frame: WindowFrame = UnspecifiedFrame,
+    bindLower: Boolean = true) {
```

Comment:
can users change this value?  seems you do not provide api to change this.
and when it will be false?  
i am thinking of remove this, since this is always true for hive sql syntax